src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -815,6 +793,32 @@ public struct Enumerator: IEnumerator<KeyValuePair<TKey,TValue>>,
             }
         }
 
+        // This is separated into a new class to save on memory by reducing
+        // fields if Keys/Values are never called.
+        // Since it shouldn't be visible to the outside world, it also serves
+        // as the SyncRoot (also rarely used) which saves an additional field.
+        // The downside of this is that Keys/Values are called, then there will
+        // be an additional object allocation the first time, plus this class
+        // needs to be initialized with Interlocked.CompareExchange.
+        private sealed class KeysAndValuesHolder
+        {
+            private readonly Dictionary<TKey, TValue> dictionary;
```

Comment:
I don't think this field is needed. `Keys` and `Values` could be methods talking a dictionary parameter.